File: src/extras-kio/libpaste.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const KFileItem = @import("libqt6").KFileItem;
const QMimeData = @import("libqt6").QMimeData;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/kio.html)
pub const KIO = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kio.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KIO,

    pub const _is_KIO = {};

    /// ### [Upstream resources](https://api.kde.org/kio.html#canPasteMimeData)
    ///
    /// ## Parameter(s):
    ///
    /// ` param1: QMimeData `
    ///
    pub fn CanPasteMimeData(param1: anytype) bool {
        comptime _ = @TypeOf(param1)._is_QMimeData;
        return qtc.KIO_CanPasteMimeData(@ptrCast(param1.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/kio.html#pasteActionText)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` param1: QMimeData `
    ///
    /// ` param2: *bool `
    ///
    /// ` param3: KFileItem `
    ///
    pub fn PasteActionText(allocator: std.mem.Allocator, param1: anytype, param2: *bool, param3: anytype) []const u8 {
        comptime _ = @TypeOf(param1)._is_QMimeData;
        comptime _ = @TypeOf(param3)._is_KFileItem;
        var _str = qtc.KIO_PasteActionText(@ptrCast(param1.ptr), @ptrCast(param2), @ptrCast(param3.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio.PasteActionText: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### [Upstream resources](https://api.kde.org/kio.html#setClipboardDataCut)
    ///
    /// ## Parameter(s):
    ///
    /// ` param1: QMimeData `
    ///
    /// ` param2: bool `
    ///
    pub fn SetClipboardDataCut(param1: anytype, param2: bool) void {
        comptime _ = @TypeOf(param1)._is_QMimeData;
        qtc.KIO_SetClipboardDataCut(@ptrCast(param1.ptr), param2);
    }

    /// ### [Upstream resources](https://api.kde.org/kio.html#isClipboardDataCut)
    ///
    /// ## Parameter(s):
    ///
    /// ` param1: QMimeData `
    ///
    pub fn IsClipboardDataCut(param1: anytype) bool {
        comptime _ = @TypeOf(param1)._is_QMimeData;
        return qtc.KIO_IsClipboardDataCut(@ptrCast(param1.ptr));
    }
};

File: src/extras-kio/libpaste.cpp
#include <KFileItem>
#include <QMimeData>
#include <QString>
#include <paste.h>
#include "libpaste.h"
#include "libpaste.hxx"

bool KIO_CanPasteMimeData(const QMimeData* data) {
    return KIO::canPasteMimeData(data);
}

libqt_string KIO_PasteActionText(const QMimeData* mimeData, bool* enable, const KFileItem* destItem) {
    auto _ret = KIO::pasteActionText(mimeData, enable, *destItem);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KIO_SetClipboardDataCut(QMimeData* mimeData, bool cut) {
    KIO::setClipboardDataCut(mimeData, cut);
}

bool KIO_IsClipboardDataCut(const QMimeData* mimeData) {
    return KIO::isClipboardDataCut(mimeData);
}
